libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -2351,6 +2351,44 @@ string YulUtilFunctions::copyArrayFromStorageToMemoryFunction(ArrayType const& _
 	});
 }
 
+string YulUtilFunctions::bytesConcatFunction(vector<Type const*> const& _argumentTypes)
+{
+	string functionName = "bytes_concat";
+	size_t totalParams = 0;
+	for (Type const* argumentType: _argumentTypes)
+	{
+		solAssert(
+			argumentType->isImplicitlyConvertibleTo(*TypeProvider::bytesMemory()) ||
+			argumentType->isImplicitlyConvertibleTo(*TypeProvider::fixedBytes(32)),
+			""
+		);
+		totalParams += argumentType->sizeOnStack();
+		functionName += "_" + argumentType->identifier();
+	}
+
+	return m_functionCollector.createFunction(functionName, [&]() {
+		Whiskers templ(R"(
+			function <functionName>(<parameters>) -> outPtr {
+				outPtr := <allocateUnbounded>()
+				let dataStart := add(outPtr, 0x20)
+				let dataEnd := <encodePacked>(dataStart<?+parameters>, <parameters></+parameters>)
+				mstore(outPtr, sub(dataEnd, dataStart))
+				<finalizeAllocation>(outPtr, sub(dataEnd, outPtr))
```

Comment:
Does the order of `mstore(outPtr, ...)` and `<finalizeAllocation>(outPtr, ...)` matter? There would be an earlier panic if `finalize` is done before storing `outPtr`. Probably not relevant.

Reply:
In `encodePacked` order was the same like here, I think it should be fine like it is...